pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -77,6 +76,29 @@ type ResolvedPipelineRunTask struct {
 // state of the PipelineRun.
 type PipelineRunState []*ResolvedPipelineRunTask
 
+// ResolvedPipelineRun represents the DAG of a pipeline along with its runtime state
+type ResolvedPipelineRun struct {
+
+	// Dag is the graph of all the pipeline tasks in the main pipeline
+	Dag *dag.Graph
+
+	// Final is the graph of all the final tasks executed once the main pipeline is complete
+	Final *dag.Graph
+
+	// State represents the state of all pipeline tasks in the pipeline
+	DagState PipelineRunState
+
+	// FinalState represents the state of all the final tasks in the pipeline
+	FinalState PipelineRunState
+
+	// WontRun is a hash of PipelineTask names that won't be executed because they are
+	// not reachable via the DAG due to the pipeline state, or had failed conditions
+	WontRun map[string]struct{}
+}
+
+// IsDone returns true if a PipelineTask was successful or failed all retries
+// NOTE: a PipelineTask that has been cancelled is not always treated as "done",
+// the condition must be checked separately
 func (t ResolvedPipelineRunTask) IsDone() (isDone bool) {
```

Comment:
named return value? 😉

Reply:
I still have to rebase this on top of your last PRs

Reply:
😿